### PR TITLE
Live Encode - Updated Presets [0 to 7]

### PR DESCRIPTION
**Pull request by hassount**
+ Adaptive Depth Partitioning (ADP) [ADAPTIVE_DEPTH_PARTITIONING]
+ add paeth pred mode c for 10bit [ENABLE_PAETH]
+ remove duplicate candidates at MD [REMOVED_DUPLICATE_INTER]
+ fix valgrind issues
+ fix crash in intra prediction due to unaligned buffers [h_predictor_16x16]

Preset Tuning
+ remove ENCODER_MODE_CLEANUP macro and multi-mode macros [Cleanup]
+ new presets 5,6,7
+ adjust all presets 0 to 7 [1080p30 on xeon 8180]

Authors:
@hguermaz
@okhlif
@arthurkkhi
@OmranAbdelkafi
@hassount

Signed-off-by: htmar <hassene.tmar@intel.com>

### CODE REVIEW

**Comment by EwoutH** [comment]
Quick bench (1080p 8-bit 4:2:0, 240 frames, Xeon E5-2680v4):

| Speed (fps) | a6c1f81 | 86766e5 |
|------------:|--------:|--------:|
| -enc-mode 7 |         |  13,378 |
| -enc-mode 6 |         |   9,040 |
| -enc-mode 5 |         |   7,046 |
| -enc-mode 4 |   3,325 |   3,761 |
| -enc-mode 3 |   3,631 |   1,717 |
| -enc-mode 2 |   1,049 |   0,591 |
| -enc-mode 1 |   0,465 |   0,413 |
| -enc-mode 0 |   0,240 |   0,258 |

The higher modes are insanely fast. The gap betweeon mode 2 and mode 3 is now quite large though.

**Comment by hassount** [comment]
@EwoutH 
Thank you for the test :) we are working on optimizing the CPU utilization of the lower presets 0,1, and 2 to make the speed gap more consistent. 

there's more to come, please let us know if you find anything wrong that we can fix while the PR is still pending

**Comment by EwoutH** [comment]
@hassount You're welcome, it's great how fast SVT-AV1 is improving.

Something I would like to test is AVX-512 performance. I don't have access to a CPU unfortunately.

**Comment by EwoutH** [comment]
I did a speed test on the Google Cloud platform, using a 96 vCPU instance. The run with 200 frames was faster then the run with 1000 frames, probably because of the lower complexity (the bitrate is also lower). Related: #98.

| Speed (fps) | -n 1000 | -n 200 |
|-------------|--------:|-------:|
| -enc-mode 7 |  18,433 | 22,545 |
| -enc-mode 6 |  13,738 | 16,248 |
| -enc-mode 5 |  11,415 | 13,258 |
| -enc-mode 4 |   7,650 |  8,559 |
| -enc-mode 3 |   2,587 |  2,973 |
| -enc-mode 2 |         |  0,874 |
| -enc-mode 1 |         |  0,511 |
| -enc-mode 0 |         |  0,317 |

- **Build:** 1fe6c26 (2019-03-02)
- **Video:** [Sintel trailer](https://media.xiph.org/) (1080p 8-bit 4:2:0)
- **System:** Google Cloud `n1-standard-96`, 96 vCPU's (Skylake), 380 GB RAM

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=262073923).

File: Source/Lib/Codec/EbEncHandle.c [hunk 1]
```diff
@@ -422,14 +421,26 @@ void SwitchToRealTime(){
...
-    fps = fps > 120 ? 120 : fps;
-    fps = fps < 24 ? 24 : fps;
-
-    return ((fps*5)>>2); // 1.25 sec worth of internal buffering
+uint32_t set_parent_pcs(EbSvtAv1EncConfiguration*   config) {
+
+    if (config){
+        uint32_t fps            = (uint32_t)((config->frame_rate > 1000) ? 
+                        config->frame_rate >> 16 : 
+                        config->frame_rate);
```
Comment (kylophone) [inline Source/Lib/Codec/EbEncHandle.c]: fps should be computed as `num / den`.

File: Source/Lib/Codec/EbEncHandle.c [hunk 2]
```diff
@@ -422,14 +421,26 @@ void SwitchToRealTime(){
...
+        fps        = fps > 120 ? 120   : fps;
+        fps        = fps < 24  ? 24    : fps; 
+        ppcs_count = MAX(min_ppcs_count, fps);
+        ppcs_count = ((ppcs_count * 5) >> 2);  // 1.25 sec worth of internal buffering
+    
+        return ppcs_count;
+    }
+    else{
+        SVT_LOG("SVT[error]: Configuration struct is corrupted\n");
+        return 60;
```
Comment (kylophone) [inline Source/Lib/Codec/EbEncHandle.c]: Use an `EbErrorType`?

File: Source/Lib/Codec/EbEncHandle.c (hunk 1 above)
Comment (hassount) [reply Source/Lib/Codec/EbEncHandle.c]: It is, line 2355 in the same file
it's also written properly to the ivf header in write_ivf_stream_header

But for the ease of use we use within the code a consolidated fps field is used

File: Source/Lib/Codec/EbEncHandle.c (hunk 2 above)
Comment (hassount) [reply Source/Lib/Codec/EbEncHandle.c]: agreed, a98c381 to address that

File: Source/Lib/Codec/EbModeDecisionConfigurationProcess.c [hunk 3]
```diff
@@ -1593,364 +1680,948 @@ void AuraDetection(
...
 }
 
+EbErrorType derive_default_segments(
+    PictureControlSet_t                *picture_control_set_ptr,
+    ModeDecisionConfigurationContext_t *context_ptr)
+{
+    EbErrorType return_error = EB_ErrorNone;
+
+#if 1
+#if 1
```
Comment (joelsole) [inline Source/Lib/Codec/EbModeDecisionConfigurationProcess.c]: Code is added under "#if 1" and "#if 0" => a defined macro would help to identify the purpose of these blocks

Comment (hassount) [reply Source/Lib/Codec/EbModeDecisionConfigurationProcess.c]: agreed, debug code
removed in 2f27461

**Comment by hassount** [comment]
@joelsole @kylophone 
Please let me know if we can proceed with this.

**Pull request merged**